Prepare genome metadata with datasets

MatBarba opened the pull request:
Instead of using a request to the ENA API, get the metadata from the JSON report from NCBI's datasets.
This simplifies the code (and the tests) since we don't need to check/mock that part.

Now we need to download the dataset file before running the script, so I've updated the Nextflow module to do so.


I've also modified the prepare part in to incorporate the "annotation" part of the genome json if there are annotations, and we can use this information to determine how many files we expect at the end of the Nextflow pipeline (3 or 6).
Also note that I've replaced `concat` by `mix`, since `concat` actually waits for the whole list of each channel before passing the next one, and mix push they as they arrive.

This change makes the `groupTuple()` work correctly: now instead of waiting for all the channels to finish, it starts as soon as it has all the files for a genome.

JAlvarezJarreta reviewed (commented):
My main concern about this approach is that it forces `datasets` to be installed for it to work (or procure a container with python, ensembl-genomio and nextflow). Since this is out of scope for the time being, what about splitting `prepare_genome_metadata.nf` in two: one where `datasets` is run (so we can use Lahcen's container), and another where `genome_metadata_prepare` is run, so the local environment is used.

MatBarba commented:
> My main concern about this approach is that it forces `datasets` to be installed for it to work (or procure a container with python, ensembl-genomio and nextflow). Since this is out of scope for the time being, what about splitting `prepare_genome_metadata.nf` in two: one where `datasets` is run (so we can use Lahcen's container), and another where `genome_metadata_prepare` is run, so the local environment is used.

Yes I was considering doing this as well since it'd make caching easier too

MatBarba commented:
I've separated the datasets download from the preparation step (and I've made it so that the datasets json files are cached, like the downloaded files)

MatBarba commented:
I have fixed a few issues, notably where I was not in fact using the correct annotation flag. I've also added a separate step so that the download of the summary file only requires an accession value.

JAlvarezJarreta commented on pipelines/nextflow/modules/genome_metadata/datasets_metadata.nf:
```diff
@@ -0,0 +1,40 @@
+        
+    shell:
+        '''
+        datasets summary genome accession !{accession} | jq '.' > ncbi_meta.json
+        if [ ! -s "ncbi_meta.json" ]; then
+            echo "No Metadata from datasets for !{accession}"
+            exit 1
+        fi
```
There are actually 2 types of "errors" from `datasets`: badly formatted accession (prints message in stderr) or accession not matching anything in NCBI (`{"total_counts": 0}`). I believe we want to consider both:
```suggestion
        datasets summary genome accession !{accession} > ncbi_meta.json
        if [ "$?" -ne 0 ]; then
            echo "Invalid or unsupported assembly accession: !{accession}"
            exit 1
        elif [[ $(jq -r '.total_count' ncbi_meta.json) -eq 0 ]]; then
            echo "No metadata returned for !{accession}"
            exit 1
        fi
```
Note that I have removed `| jq '.'` because if left it "nullifies" `$?` (makes it 0 whatever `datasets` returned).

JAlvarezJarreta commented on pipelines/nextflow/modules/genome_metadata/prepare_genome_metadata.nf:
```diff
@@ -14,23 +14,18 @@
     output_json = "genome.json"
     '''
-    genome_metadata_prepare --input_file !{input_json} --output_file !{output_json}
+    genome_metadata_prepare --input_file !{input_json} --output_file !{output_json} --ncbi_meta !{ncbi_json}
```
Why have you removed the stub?

MatBarba replied:
This part no longer needs to be mocked: the ncbi_meta generated by the datasets module already produces the mock file in its own stub

JAlvarezJarreta commented on pipelines/nextflow/modules/genome_metadata/datasets_metadata.nf:
```diff
@@ -0,0 +1,40 @@
+    tag "$accession"
+    label 'local'
+    label 'cached'
+
```
Shall we add the container here? Since `NXF_SINGULARITY_CACHEDIR` is pointing to were our current container is, it should work for us. We can wait until the next containers meeting to see how we release it to the world after that. I assume `nextflow.config` will need to be also updated to enable singularity.

MatBarba replied:
When we agree on the container yes

JAlvarezJarreta commented on pipelines/nextflow/workflows/genome_prepare/main.nf:
```diff
@@ -49,22 +51,31 @@ def meta_from_genome_json(json_path) {
+    }
 
     return [
+        // id: prod_name,
```
```suggestion
```

JAlvarezJarreta commented on src/python/ensembl/io/genomio/genome_metadata/prepare.py:
```diff
@@ -75,15 +68,15 @@ class MetadataError(Exception):
     Args:
         genome_data: Genome information of assembly, accession and annotation.
-        gff3_file: Path to GFF3 file to use as annotation source for this genome.
+        ncbi_data: Report data from datasets for this accession.
```
```suggestion
        ncbi_data: Report data from NCBI datasets.
```

JAlvarezJarreta commented on src/python/ensembl/io/genomio/genome_metadata/prepare.py:
```diff
@@ -141,95 +134,35 @@ def add_genebuild_metadata(genome_data: Dict) -> None:
     Args:
         genome_metadata: Genome information of assembly, accession and annotation.
-        base_api_url: Base API URL to fetch the taxonomy data from.
+        ncbi_data: Report data from datasets for this accession.
```
```suggestion
        ncbi_data: Report data from NCBI datasets.
```

JAlvarezJarreta commented on src/python/ensembl/io/genomio/genome_metadata/prepare.py:
```diff
@@ -239,20 +172,19 @@ def prepare_genome_metadata(
-        gff3_file: Path to GFF3 file to use as annotation source for this genome.
-        base_api_url: Base API URL to fetch the taxonomy data from.
-        mock_run: Do not call external taxonomy service.
+        ncbi_meta: JSON file from NCBI datasets for the accession.
```
```suggestion
        ncbi_meta: JSON file from NCBI datasets.
```

JAlvarezJarreta commented on src/python/ensembl/io/genomio/genome_metadata/prepare.py:
```diff
@@ -264,19 +196,13 @@ def main() -> None:
-    parser.add_argument(
-        "--base_api_url", default=DEFAULT_API_URL, help="API URL to fetch the taxonomy data from"
+    parser.add_argument_src_path(
+        "--ncbi_meta", required=True, help="JSON file from NCBI datasets for this accession."
```
I have removed the reference to `accession` in previous documentation because there is no such argument anywhere.
```suggestion
        "--ncbi_meta", required=True, help="JSON file from NCBI datasets for this genome."
```

JAlvarezJarreta commented on src/python/tests/genome_metadata/test_prepare.py:
```diff
@@ -90,13 +83,13 @@ def test_add_provider(
         json_data: JSON test file parsing fixture.
         genome_file: Genome metadata JSON file.
-        gff3_file: GFF3 file.
+        ncbi_data: Report from datasets.
```
```suggestion
        ncbi_data: Report from NCBI datasets.
```

JAlvarezJarreta commented on src/python/tests/genome_metadata/test_prepare.py:
```diff
@@ -150,195 +143,93 @@ def test_add_genebuild_metadata(
-        tmp_path: Test's unique temporary directory fixture.
-        data_dir: Module's test data directory fixture.
-        assert_files: File diff assertion fixture.
+        input_filename: Genome json input.
+        ncbi_filename: NCBI dataset json input.
+        expected_filename: Genome json input expected output.
```
```suggestion
        input_filename: Input genome JSON file.
        ncbi_filename: NCBI dataset report JSON file.
        expected_filename: Expected output genome JSON file.
```

JAlvarezJarreta reviewed (changes requested):
Quite a nice update of the codebase, showing the potential of `datasets` CLI tool :nerd_face:

JAlvarezJarreta commented on src/python/ensembl/io/genomio/genome_metadata/prepare.py:
```diff
@@ -141,95 +134,35 @@ def add_genebuild_metadata(genome_data: Dict) -> None:
-    `"scientific_name"` and `"strain"` will be fetched from the taxonomy information linked to the given
-    accession.
+def add_species_metadata(genome_metadata: Dict, ncbi_data: Dict) -> None:
+    """Adds tax_id and name from the NCBI dataset report.
```
```suggestion
    """Adds taxonomy ID, scientific name and strain (if present) from the NCBI dataset report.
```

JAlvarezJarreta reviewed (approved):
One tiny docstring update, but this is now ready to go!
